src/OT/glyf/Glyph.hh

```diff
@@ -72,12 +72,109 @@ struct Glyph
+    if (vert_aw < 0) vert_aw = 0;
+    int tsb = roundf (topSideY - yMax);
+    plan->vmtx_map->set (new_gid, hb_pair (vert_aw, tsb));
+  }
+
+  bool compile_header_bytes (const hb_subset_plan_t *plan,
+                             const contour_point_vector_t &all_points,
+                             hb_bytes_t &dest_bytes /* OUT */) const
+  {
+    if (all_points.length == 4) //Empty glyph
+    {
+      dest_bytes = hb_bytes_t ();
+      return true;
+    }
+
+    GlyphHeader *glyph_header = (GlyphHeader *) hb_calloc (1, GlyphHeader::static_size);
+    if (unlikely (!glyph_header)) return false;
+
+    int xMin, xMax;
+    xMin = xMax = roundf (all_points[0].x);
+
+    int yMin, yMax;
+    yMin = yMax = roundf (all_points[0].y);
+
+    for (unsigned i = 1; i < all_points.length - 4; i++)
+    {
+      float rounded_x = roundf (all_points[i].x);
+      float rounded_y = roundf (all_points[i].y);
+      xMin = hb_min (xMin, rounded_x);
+      xMax = hb_max (xMax, rounded_x);
+      yMin = hb_min (yMin, rounded_y);
+      yMax = hb_max (yMax, rounded_y);
+    }
+
+    update_mtx (plan, xMin, yMax, all_points);
+
+    glyph_header->numberOfContours = header->numberOfContours;
+    glyph_header->xMin = xMin;
+    glyph_header->yMin = yMin;
+    glyph_header->xMax = xMax;
+    glyph_header->yMax = yMax;
+
+    dest_bytes = hb_bytes_t ((const char *)glyph_header, GlyphHeader::static_size);
+    return true;
+  }
+
+  bool compile_bytes_with_deltas (const hb_subset_plan_t *plan,
+                                  hb_font_t *font,
+                                  const glyf_accelerator_t &glyf,
+                                  hb_bytes_t &dest_start,  /* IN/OUT */
+                                  hb_bytes_t &dest_end /* OUT */) const
+  {
+    contour_point_vector_t all_points, deltas;
+    get_points (font, glyf, all_points, deltas);
+
+    switch (type) {
+    case COMPOSITE:
+      if (!CompositeGlyph (*header, bytes).compile_bytes_with_deltas (dest_start,
+                                                                      deltas,
+                                                                      dest_end))
+        return false;
+      break;
+    case SIMPLE:
+      if (!SimpleGlyph (*header, bytes).compile_bytes_with_deltas (all_points,
+                                                                   plan->flags & HB_SUBSET_FLAGS_NO_HINTING,
+                                                                   dest_end))
+        return false;
+      break;
+    default:
+      //no need to compile empty glyph (.notdef)
+      return true;
+    }
+
+    return compile_header_bytes (plan, all_points, dest_start);
+  }
+
   /* Note: Recursively calls itself.
    * all_points includes phantom points
    */
   template <typename accelerator_t>
   bool get_points (hb_font_t *font, const accelerator_t &glyf_accelerator,
 		   contour_point_vector_t &all_points /* OUT */,
+		   contour_point_vector_t &deltas, /* OUT */
```

Comment:
I don't understand what the deltas here is. This function is very performance critical.

Reply:
This slows down glyph-drawing by up to 10%. I'd rather not do that. Test with:
```
build/perf/benchmark-font --benchmark_filter=glyph_shape.*hb
```

Reply:
get_points() gets a flattened list of all contour points. When instancing composite glyphs, we actually only need the deltas for x/y offset, no need to do recursive call on child glyphs. So I added code to fill these deltas only when level is 0. If that slows down glyph-drawing, I can make another function to get points  for Composite glyphs only.

Reply:
uh... I think for recalculating boundaries, we still need to expand child glyphs' points. So I would probably just add a new function for subset use, which is mainly a copy of get_points()

Reply:
get_points() is a pretty sizable method so we should try and find a way to avoid having to duplicate the code in it.

To avoid duplicating all of the code from get_points() while maintaining performance you could try templating the method (ie. maybe like a boolean template parameter that says wether or not to get the deltas). That way in the case where you don't need to get deltas the templated function should compile to equivalent as the current version.

Reply:
I agree. Either template or just passing a nullptr for deltas.

Reply:
Done